File: android/src/main/java/com/margelo/nitro/audiorecorderplayer/HybridNativeWebSocket.kt
package com.margelo.nitro.sound

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.util.concurrent.TimeUnit

class HybridNativeWebSocket : HybridNativeWebSocketSpec() {

    private var ws: WebSocket? = null
    private var client: OkHttpClient? = null

    @Volatile
    private var _isConnected = false

    private var onOpenListener: (() -> Unit)? = null
    private var onTextMessageListener: ((String) -> Unit)? = null
    private var onCloseListener: ((Double, String) -> Unit)? = null
    private var onErrorListener: ((String) -> Unit)? = null

    override val isConnected: Boolean
        get() = _isConnected

    override val memorySize: Long
        get() = 0

    // ─── Lifecycle ──────────────────────────────────────────────────

    override fun connect(url: String): Promise<Unit> {
        val promise = Promise<Unit>()

        cleanupConnection()

        val httpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .pingInterval(25, TimeUnit.SECONDS)
            .build()

        client = httpClient

        val request = Request.Builder()
            .url(url)
            .build()

        var resolved = false

        ws = httpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _isConnected = true
                onOpenListener?.invoke()
                if (!resolved) {
                    resolved = true
                    promise.resolve(Unit)
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                onTextMessageListener?.invoke(text)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                val text = bytes.utf8()
                onTextMessageListener?.invoke(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                _isConnected = false
                webSocket.close(code, reason)
                onCloseListener?.invoke(code.toDouble(), reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _isConnected = false
                onCloseListener?.invoke(code.toDouble(), reason)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _isConnected = false
                val errorMsg = t.message ?: "Unknown WebSocket error"
                onErrorListener?.invoke(errorMsg)
                onCloseListener?.invoke(1006.0, errorMsg)
                if (!resolved) {
                    resolved = true
                    promise.reject(Exception(errorMsg))
                }
            }
        })

        return promise
    }

    override fun disconnect(code: Double, reason: String) {
        _isConnected = false
        ws?.close(code.toInt(), reason)
    }

    // ─── Send ───────────────────────────────────────────────────────

    override fun sendBinary(data: ArrayBuffer) {
        if (!_isConnected) return
        val buffer = data.getBuffer(false)
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        val byteString = bytes.toByteString()
        val success = ws?.send(byteString) ?: false
        if (!success) {
            onErrorListener?.invoke("sendBinary failed: WebSocket send returned false")
        }
    }

    override fun sendText(text: String) {
        if (!_isConnected) return
        val success = ws?.send(text) ?: false
        if (!success) {
            onErrorListener?.invoke("sendText failed: WebSocket send returned false")
        }
    }

    override fun sendPing() {
        // OkHttp handles ping/pong automatically via pingInterval.
        // Manual ping is not exposed, but the 25s auto-ping covers this.
    }

    // ─── Listeners ──────────────────────────────────────────────────

    override fun setOnOpenListener(callback: () -> Unit) {
        onOpenListener = callback
    }

    override fun setOnTextMessageListener(callback: (String) -> Unit) {
        onTextMessageListener = callback
    }

    override fun setOnCloseListener(callback: (Double, String) -> Unit) {
        onCloseListener = callback
    }

    override fun setOnErrorListener(callback: (String) -> Unit) {
        onErrorListener = callback
    }

    override fun removeAllListeners() {
        onOpenListener = null
        onTextMessageListener = null
        onCloseListener = null
        onErrorListener = null
    }

    // ─── Cleanup ────────────────────────────────────────────────────

    private fun cleanupConnection() {
        _isConnected = false
        ws?.cancel()
        ws = null
        client?.dispatcher?.executorService?.shutdown()
        client = null
    }

    protected fun finalize() {
        cleanupConnection()
    }
}

File: ios/HybridNativeWebSocket.swift
import Foundation
import NitroModules

final class HybridNativeWebSocket: HybridNativeWebSocketSpec_base, HybridNativeWebSocketSpec_protocol {

    // MARK: - Properties

    private var session: URLSession?
    private var task: URLSessionWebSocketTask?
    private let queue = DispatchQueue(label: "com.nitrosound.websocket", qos: .userInitiated)
    private var operationQueue: OperationQueue = {
        let q = OperationQueue()
        q.name = "NativeWebSocket"
        q.maxConcurrentOperationCount = 1
        return q
    }()

    private var onOpenListener: (() -> Void)?
    private var onTextMessageListener: ((String) -> Void)?
    private var onCloseListener: ((Double, String) -> Void)?
    private var onErrorListener: ((String) -> Void)?

    private var _isConnected = false
    private var isReceiving = false

    var isConnected: Bool { _isConnected }

    var memorySize: Int { 0 }

    // MARK: - Lifecycle

    func connect(url: String) throws -> Promise<Void> {
        let promise = Promise<Void>()

        queue.async { [weak self] in
            guard let self = self else {
                promise.reject(withError: RuntimeError.error(withMessage: "WebSocket deallocated"))
                return
            }

            self.cleanupConnection()

            guard let wsUrl = URL(string: url) else {
                promise.reject(withError: RuntimeError.error(withMessage: "Invalid WebSocket URL"))
                return
            }

            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = 10
            config.waitsForConnectivity = false

            self.session = URLSession(
                configuration: config,
                delegate: nil,
                delegateQueue: self.operationQueue
            )

            self.task = self.session?.webSocketTask(with: wsUrl)

            var resolved = false

            self.task?.resume()
            self._isConnected = true
            self.isReceiving = true
            self.startReceiveLoop(connectPromise: promise, resolved: &resolved)

            if !resolved {
                resolved = true
                promise.resolve(withResult: ())
            }
        }

        return promise
    }

    func disconnect(code: Double, reason: String) throws {
        queue.async { [weak self] in
            guard let self = self else { return }
            let closeCode = URLSessionWebSocketTask.CloseCode(rawValue: Int(code)) ?? .normalClosure
            self.task?.cancel(with: closeCode, reason: reason.data(using: .utf8))
            self._isConnected = false
            self.isReceiving = false
        }
    }

    // MARK: - Send

    func sendBinary(data: ArrayBuffer) throws {
        guard _isConnected, let task = task else { return }

        let nsData = data.toData()
        let message = URLSessionWebSocketTask.Message.data(nsData)

        task.send(message) { [weak self] error in
            if let error = error {
                self?.queue.async {
                    self?.onErrorListener?("sendBinary failed: \(error.localizedDescription)")
                }
            }
        }
    }

    func sendText(text: String) throws {
        guard _isConnected, let task = task else { return }

        let message = URLSessionWebSocketTask.Message.string(text)

        task.send(message) { [weak self] error in
            if let error = error {
                self?.queue.async {
                    self?.onErrorListener?("sendText failed: \(error.localizedDescription)")
                }
            }
        }
    }

    func sendPing() throws {
        task?.sendPing { [weak self] error in
            if let error = error {
                self?.queue.async {
                    self?.onErrorListener?("ping failed: \(error.localizedDescription)")
                }
            }
        }
    }

    // MARK: - Listeners

    func setOnOpenListener(callback: @escaping () -> Void) throws {
        onOpenListener = callback
    }

    func setOnTextMessageListener(callback: @escaping (String) -> Void) throws {
        onTextMessageListener = callback
    }

    func setOnCloseListener(callback: @escaping (Double, String) -> Void) throws {
        onCloseListener = callback
    }

    func setOnErrorListener(callback: @escaping (String) -> Void) throws {
        onErrorListener = callback
    }

    func removeAllListeners() throws {
        onOpenListener = nil
        onTextMessageListener = nil
        onCloseListener = nil
        onErrorListener = nil
    }

    // MARK: - Receive Loop

    private func startReceiveLoop(connectPromise: Promise<Void>? = nil, resolved: inout Bool) {
        scheduleReceive()

        if let listener = onOpenListener {
            listener()
        }
    }

    private func scheduleReceive() {
        guard isReceiving, let task = task else { return }

        task.receive { [weak self] result in
            guard let self = self, self.isReceiving else { return }

            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self.onTextMessageListener?(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self.onTextMessageListener?(text)
                    }
                @unknown default:
                    break
                }
                self.scheduleReceive()

            case .failure(let error):
                self._isConnected = false
                self.isReceiving = false

                let nsError = error as NSError
                if nsError.code == 57 || // Socket is not connected
                   nsError.domain == NSPOSIXErrorDomain {
                    self.onCloseListener?(
                        Double(URLSessionWebSocketTask.CloseCode.abnormalClosure.rawValue),
                        error.localizedDescription
                    )
                } else {
                    self.onErrorListener?(error.localizedDescription)
                    self.onCloseListener?(
                        Double(URLSessionWebSocketTask.CloseCode.abnormalClosure.rawValue),
                        error.localizedDescription
                    )
                }
            }
        }
    }

    // MARK: - Cleanup

    private func cleanupConnection() {
        isReceiving = false
        _isConnected = false
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        session?.invalidateAndCancel()
        session = nil
    }

    deinit {
        cleanupConnection()
        onOpenListener = nil
        onTextMessageListener = nil
        onCloseListener = nil
        onErrorListener = nil
    }
}

// MARK: - ArrayBuffer Extension

private extension ArrayBuffer {
    func toData() -> Data {
        let ptr = data
        let size = Int(self.size)
        return Data(bytes: ptr, count: size)
    }
}
